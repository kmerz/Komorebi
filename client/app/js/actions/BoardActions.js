/*jshint esversion: 6 */
import AppDispatcher from '../dispatcher/AppDispatcher';
var BoardActions = {
  fetchBoard: () => {
    AppDispatcher.dispatch({
      actionType: "FETCH_BOARD",
    });
  },
  fetchBoards: () => {
    AppDispatcher.dispatch({
      actionType: "FETCH_BOARDS",
    });
  },
  fetchStories: () =>  {
    AppDispatcher.dispatch({
      actionType: "FETCH_STORIES",
    });
  },
  fetchTasks: () => {
    AppDispatcher.dispatch({
      actionType: "FETCH_TASKS",
    });
  },
  fetchAll: () =>  {
    AppDispatcher.dispatch({
      actionType: "FETCH_ALL",
    });
  },
  showTaskDialog: () =>  {
    AppDispatcher.dispatch({
      actionType: "SHOW_TASK_DIALOG"
    });
  },
  closeTaskDialog: (reload) =>  {
    AppDispatcher.dispatch({
      actionType: "CLOSE_TASK_DIALOG",
      reload: reload
    });
    AppDispatcher.dispatch({
      actionType: "REMOVE_TASK_ERRORS"
    });
  },
  openStoryShowDialog: (story_id) =>  {
    AppDispatcher.dispatch({
      actionType: "OPEN_STORY_SHOW_DIALOG",
      story_id: story_id
    });
  },
  closeStoryShowDialog: () =>  {
    AppDispatcher.dispatch({
      actionType: "CLOSE_STORY_SHOW_DIALOG",
    });
  },
  openStoryEditDialog: () =>  {
    AppDispatcher.dispatch({
      actionType: "OPEN_STORY_EDIT_DIALOG"
    });
  },
  openStoryFromIssueEditDialog: () =>  {
    AppDispatcher.dispatch({
      actionType: "OPEN_STORY_FROM_ISSUE_EDIT_DIALOG"
    });
  },
  closeStoryEditDialog: (reload) =>  {
    AppDispatcher.dispatch({
      actionType: "CLOSE_STORY_EDIT_DIALOG",
      reload: reload
    });
  },
  closeStoryFromIssueEditDialog: (reload) =>  {
    AppDispatcher.dispatch({
      actionType: "CLOSE_STORY_FROM_ISSUE_EDIT_DIALOG",
      reload: reload
    });
  },
  showColumnDialog: () =>  {
    AppDispatcher.dispatch({
      actionType: "SHOW_COLUMN_DIALOG"
    });
  },
  closeColumnDialog: (reload) =>  {
    AppDispatcher.dispatch({
      actionType: "CLOSE_COLUMN_DIALOG",
      reload: reload
    });
  },
  showTasksForStoryId: (story_id) =>  {
    AppDispatcher.dispatch({
      actionType: "SHOW_TASKS_FOR_STORY_ID",
      story_id: story_id
    });
  },
  updateTask: (data) => {
    AppDispatcher.dispatch({
      actionType: "UPDATE_TASK",
      data: data
    });
  },
  openBoardDialog: () => {
    AppDispatcher.dispatch({
      actionType: "OPEN_BOARD_DIALOG"
    });
  },
  closeBoardDialog: () => {
    AppDispatcher.dispatch({
      actionType: "CLOSE_BOARD_DIALOG"
    });
  },
  addBoard: (data) => {
    AppDispatcher.dispatch({
      actionType: "ADD_BOARD",
      data: data
    });
  },
  addColumn: (data) => {
    AppDispatcher.dispatch({
      actionType: "ADD_COLUMN",
      data: data
    });
  },
  addTask: (data) => {
    AppDispatcher.dispatch({
      actionType: "ADD_TASK",
      data: data
    });
  },
  updateSelectedStoryId: (story_id) => {
    AppDispatcher.dispatch({
      actionType: "UPDATE_SELECTED_STORY_ID",
      id: story_id
    });
  },
  addStory: (data) => {
    AppDispatcher.dispatch({
      actionType: "ADD_STORY",
      data: data
    });
  },
  initBoard: () => {
    AppDispatcher.dispatch({
      actionType: "INIT_BOARD"
    });
  },
  closeMessage: () => {
    AppDispatcher.dispatch({
      actionType: "CLOSE_MESSAGE"
    });
  }
};
module.exports = BoardActions;
