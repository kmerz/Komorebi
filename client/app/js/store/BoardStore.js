/*jshint esversion: 6 */
import AppDispatcher from '../dispatcher/AppDispatcher';
import EventEmitter from 'events';
import assign from 'object-assign';
import Ajax from  'basic-ajax';

var board_id = null;
var board_title = "";
var columns = [];
var stories = [];
var tasks = [];

var menu_open = false;
var column_dialog_open = false;
var story_edit_dialog_open = false;
var story_show_dialog_open = false;
var task_dialog_open = false;

var story_show_id = null;

var CHANGE_EVENT = 'change';

var BoardStore = assign({}, EventEmitter.prototype, {
  getBoardId: function () {
    return board_id;
  },
  getBoardTitle: function () {
    return board_title;
  },
  getColumns: function() {
    return columns;
  },
  getStories: function() {
    return stories;
  },
  getTasks: function() {
    return tasks;
  },
  getTaskByStoryId: function(story_id) {
    return tasks.reduce((acc, task) => {
      if (task.story_id == story_id) {
        acc.push(task);
      }
      return acc;
    }, []);
  },
  getMenuOpen: () => {
    return menu_open;
  },
  getColumnDialogOpen: () => {
    return column_dialog_open;
  },
  getStoryEditDialogOpen: () => {
    return story_edit_dialog_open;
  },
  getTaskDialogOpen: () => {
    return task_dialog_open;
  },
  getFirstColumn: () => {
    return columns.find((column) => { return column.position === 0; });
  },
  getStoryShowDialogOpen: () => {
    return story_show_dialog_open;
  },
  getStoryShowId: () => {
    return story_show_id;
  },
  emitChange: function() {
    this.emit(CHANGE_EVENT);
  },
  addChangeListener: function(callback) {
    this.on(CHANGE_EVENT, callback);
  },
  removeChangeListener: function(callback) {
    this.removeListener(CHANGE_EVENT, callback);
  }
});

var fetchBoard = () => {
  board_id = null;
  board_title = "";
  columns = [];
  return Ajax.get(window.location.pathname, {"Accept": "application/json"}).then(response => {
    if(response.status == 200) {
      var board = JSON.parse(response.responseText);
      board_id = board.id;
      board_title = board.name;
      // should get columns over action and ajax
      columns = board.columns;
    }
  });
};

var fetchStories = () => {
  stories = [];
  return Ajax.get(window.location.pathname + "/stories", {"Accept": "application/json"}).then(response => {
    if(response.status == 200) {
      stories = JSON.parse(response.responseText);
    }
  });
};

var fetchTasks = () => {
  tasks = [];
  return new Promise((resolve, reject) => {
    let count = stories.length;
    stories.forEach(function(story) {
      Ajax.get(`/stories/${story.id}/tasks`,
        {"Accept": "application/json"}).then(response => {
        if(response.status == 200) {
          let fetched_tasks = JSON.parse(response.responseText);
          Array.prototype.push.apply(tasks, fetched_tasks);
          count--;
          if(count === 0) {
            resolve();
          }
        }
      });
    });
  });
};

var fetchAll = () => {
  return new Promise((resolve, reject) => {
    fetchBoard();
    fetchStories().then(fetchTasks).then(resolve);
  });
};

AppDispatcher.register(function(action) {
  switch(action.actionType) {
    case "FETCH_ALL":
      fetchAll().
        then(() => {BoardStore.emitChange();});
      break;
    case "FETCH_BOARD":
      fetchBoard().
        then(() => {BoardStore.emitChange();});
      break;
    case "FETCH_STORIES":
      fetchStories().
        then(() => {BoardStore.emitChange();});
      break;
    case "FETCH_TASKS":
      fetchTasks().
        then(() => {BoardStore.emitChange();});
      break;
    case "OPEN_STORY_SHOW_DIALOG":
      story_show_dialog_open = true;
      story_show_id = action.story_id;
      BoardStore.emitChange();
      break;
    case "CLOSE_STORY_SHOW_DIALOG":
      story_show_dialog_open = false;
      BoardStore.emitChange();
      break;
    case "OPEN_STORY_EDIT_DIALOG":
      story_edit_dialog_open = true;
      BoardStore.emitChange();
      break;
    case "CLOSE_STORY_EDIT_DIALOG":
      story_edit_dialog_open = false;
      if (action.reload) {
        fetchAll().then(() => {BoardStore.emitChange();});
      } else {
        BoardStore.emitChange();
      }
      break;
    case "SHOW_TASK_DIALOG":
      task_dialog_open = true;
      BoardStore.emitChange();
      break;
    case "CLOSE_TASK_DIALOG":
      task_dialog_open = false;
      if (action.reload) {
        fetchAll().then(() => {BoardStore.emitChange();});
      } else {
        BoardStore.emitChange();
      }
      break;
    case "SHOW_COLUMN_DIALOG":
      column_dialog_open = true;
      BoardStore.emitChange();
      break;
    case "CLOSE_COLUMN_DIALOG":
      column_dialog_open = false;
      if (action.reload) {
        fetchAll().then(() => {BoardStore.emitChange();});
      } else {
        BoardStore.emitChange();
      }
      break;
    default:
      // no op
  }
});
module.exports = BoardStore;
