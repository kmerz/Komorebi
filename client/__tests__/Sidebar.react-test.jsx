import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import { mount } from 'enzyme';

import injectTapEventPlugin from 'react-tap-event-plugin';
import ReactTestUtils from 'react-addons-test-utils';
import ReactDOM from 'react-dom';
import Sidebar from '../app/js/sidebar';
import BoardStore from '../app/js/store/BoardStore';
import BoardActions from '../app/js/actions/BoardActions';

injectTapEventPlugin();

const test_data = {
  story1: {"id":1,"name":"Story1","updated_at":1468575129729239946,"desc":"Test","points":4,"priority":1,"requirements":"","column_id":1},
};

describe('<Sidebar />', () => {
  BoardStore.__Rewire__('stories', [test_data.story1]);
  it('Should render Sidebar', () => {
    const wrapper = shallow(
      <Sidebar />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should update active story', () => {
    const wrapper = mount(
      <Sidebar />
    );
    expect(wrapper.state().selected_stories).toEqual([]);
    BoardActions.showTasksForStoryId(1);
    expect(wrapper.state().selected_stories).toEqual([1]);
  });
});
