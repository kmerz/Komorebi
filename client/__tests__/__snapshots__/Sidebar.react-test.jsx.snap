exports[`<Sidebar /> Should render Sidebar 1`] = `
<Col
  bsClass="col"
  componentClass="div"
  lg={2}
  md={12}
  xs={12}>
  <ListGroup
    bsClass="list-group"
    className="sidebar">
    <StoryCard
      active={false}
      name="Story1"
      story_id={1} />
  </ListGroup>
</Col>
`;
